sympy/functions/special/spherical_harmonics.py

```diff
@@ -179,19 +180,25 @@ def fdiff(self, argindex=4):
             raise ArgumentIndexError(self, argindex)
 
     def _eval_rewrite_as_polynomial(self, n, m, theta, phi, **kwargs):
-        # TODO: Make sure n \in N
-        # TODO: Assert |m| <= n ortherwise we should return 0
-        return self.expand(func=True)
+        if n in S.Naturals:
+            c = is_le(Abs(m), n)
+            if c == True:
+                return self.expand(func=True)
+            elif c == False:
+                return S.Zero
 
     def _eval_rewrite_as_sin(self, n, m, theta, phi, **kwargs):
         return self.rewrite(cos)
 
     def _eval_rewrite_as_cos(self, n, m, theta, phi, **kwargs):
         # This method can be expensive due to extensive use of simplification!
         from sympy.simplify import simplify, trigsimp
-        # TODO: Make sure n \in N
-        # TODO: Assert |m| <= n ortherwise we should return 0
-        term = simplify(self.expand(func=True))
+        if n in S.Naturals:
+            c = is_le(Abs(m), n)
+            if c == True:
+                term = simplify(self.expand(func=True))
+            elif c == False:
+                term = S.Zero
         # We can do this because of the range of theta
```

Comment:
I think this part should only be done in the if-clause. 

Some tests are needed which would detect e.g. this strange behavior.

Reply:
I'm a bit confused with what kind of tests to add. Could you help me with it?